DataFormats/MemoryResources/include/MemoryResources/MemoryResources.h

```diff
@@ -136,13 +89,19 @@ class MessageResource : public FairMQMemoryResource
   FairMQMemoryResource* mUpstream{ nullptr };
   size_t mMessageSize{ 0 };
   void* mMessageData{ nullptr };
+  bool initialImport{ true };
 
   void* do_allocate(std::size_t bytes, std::size_t alignment) override
   {
-    if (bytes > mMessageSize) {
-      throw std::bad_alloc();
+    if (initialImport) {
+      if (bytes > mMessageSize) {
+        throw std::bad_alloc();
+      }
+      initialImport = false;
+      return mMessageData;
+    } else {
+      return mUpstream->allocate(bytes, alignment);
```

Comment:
I'm trying to understand the de-allocation. In `de_allocate` only the initial message is handled, regardless of the parameters. It seems that `do_alloc` can be called multiple times, the question is whether to update `mMessageData` to keep track of it. To me it looks like, `do_dealloc` must be adapted.

Reply:
indeed, fixed. thanks.

Reply:
@matthiasrichter, I could also just call mUpstream->deallocate(p, size), since the standard says it is UB if you try to deallocate something that is not yours... an exception seems easier to debug if need be...